consensus/consensus.go

```diff
@@ -585,11 +622,68 @@ func (consensus *Consensus) Prepare(chain consensus_engine.ChainReader, header *
 	return nil
 }
 
-// AccumulateRewards credits the coinbase of the given block with the mining
+// accumulateRewards credits the coinbase of the given block with the mining
 // reward. The total reward consists of the static block reward and rewards for
 // included uncles. The coinbase of each uncle block is also rewarded.
-func accumulateRewards(config *params.ChainConfig, state *state.DB, header *types.Header) {
-	// TODO: implement mining rewards
+func (consensus *Consensus) accumulateRewards(
+	config *params.ChainConfig, state *state.DB, header *types.Header,
+) {
+	if header.ParentHash == (common.Hash{}) {
+		// This block is a genesis block,
+		// without a parent block whose signer to reward.
+		return
+	}
+	if consensus.ChainReader == nil {
+		utils.GetLogInstance().Error("accumulateRewards: ChainReader is nil")
+		return
+	}
+	parent := consensus.ChainReader.GetBlockByHash(header.ParentHash)
+	if parent == nil {
+		utils.GetLogInstance().Error(
+			"accumulateRewards: cannot retrieve parent block",
+			"curHash", header.Hash().Hex(),
+			"parentHash", header.ParentHash.Hex(),
+		)
+		return
+	}
+	mask, err := bls_cosi.NewMask(consensus.PublicKeys, consensus.PubKey)
+	if err != nil {
+		utils.GetLogInstance().Error(
+			"accumulateRewards: cannot create group sig mask", "error", err)
+		return
+	}
+	parentSigners := parent.Header().CommitBitmap
+	utils.GetLogInstance().Debug("accumulateRewards: setting group sig mask",
+		"destLen", mask.Len(), "srcLen", len(parentSigners))
+	if err := mask.SetMask(parentSigners); err != nil {
+		utils.GetLogInstance().Error(
+			"accumulateRewards: cannot set group sig mask bits", "error", err)
+		return
+	}
+	numSigs := 0
+	numAccounts := 0
+	totalAmount := big.NewInt(0)
+	for _, key := range mask.GetPubKeyFromMask(true) {
+		numSigs++
```

Comment:
numSigs is just the length of mask.GetPubKeyFromMask, right? it is not used, why do we need a variable here?

Reply:
Good point, will fix.

Reply:
Fixed in commit 715a8ffc6d64b4d621bbfcf9d492cb78073511da.